README
~/dev/py/fieldz/README

2012-08-17

Think about performance of fieldz when handling large files: we want to put 
U leaf nodes on the wire without repeated copying and certainly without any 
sort of bit-shifting.  Typically large files will be read block by block 
into a fixed-length buffer (presumably N*4096) and then put on the wire as 
soon as available - block by block.  Reading from the wire will follow
a similar pattern.  In either case, whether writing or reading large files, 
a data packet will typically have a packed, variegated header followed by a 
long 'lenPlus', a varint len followed by that many raw bytes.

Obvious possibility: replace the data from the file with a content key.  
The notion might be that there is always at least one U associated with 
every connection.  U transfers on the wire look like 
    PUT         + key + len + byte[len]
    PUTI + dest + key + len + byte[len]    
where PUT means 'write to the machine at the far end of the 
connection' and PUTI means write to node 'dest' through the
machine at the far end of the connection.  

key and dest may be 160 or 256 bit values (SHA1 or SHA3 digests) or they 
may be much shorter varints, indexes into caches shared by the nodes at 
the ends of the connection.  The receiver knows from the length of the
encoding (of key or dest) whether it represents the actual value or an
index into a cache of such values.

Note the possible complementary 
    GET         + key
    GETI + src  + key
Here 'key' is resolved at the far end; if we knew its length and value, 
it would be in our cache and we would not be asking for it.  Similarly 
where the operation is PUT or PUTI 'key' must represent a datum in U, 
so the sender necessarily knows len and the value, what's in the byte array.

Why are these considerations at the fieldz level?  For the same reason
that varints are.  A varint is (usually) an economical way to represent
a value on the wire.  The fieldz package uses a set of rules to convert
between the varint encoding and raw uint values.  The proposal is to add
content keys to the set of encoding tools available at the fieldz level.

In the PUTI example above, 'dest' is either a varint or a nodeID.  If
it is a varint, the receiver either has it in its cache or it doesn't.
If it doesn't, the receiver can ask for its binding, what it represents.  
Otherwise the value it binds to is a key, the nodeID of the destination.
In the proposed scheme, this can be used to fetch the destination's 
public key from a local U.  That is
    varint -> cache index -> nodeID -> public key
Each of these refers to the same thing, to the destination node (and as
you go from left to right, each is larger).

Hmmmm.  I think that two different ideas are being commingled here.
First, fieldz needs to be able to put large data files on the wire 
and get them back off the wire efficiently.  Data in such files must
not be copied unless absolutely necessary.  It certainly must not be
bit-shifted.  Its integrity must be assured, and the use of longer
content keys (160- or 256-bit digests) is ideal for this.

At another level we can use caches and content keys to actually avoid 
transmitting data and to shorten what data is being transmitted.

We also need a layer just above fieldz which converts primitives
(varints, 32- and 64-bit fixed-length values, variable-length 
strings) to a richer collection of types.  It might make sense to
put content keys at this level.

Zig-zag encoding is naturally treated in just this way.  We decide 
that certain fields in a message are going to come from a range of
small positive and negative values.  So we zig-zag encode the value,
mapping it into a larger range of non-negative values, and then 
encode it as a varint.  There is no reason for the fieldz-level logic
to know about the zig-zag encoding.

Encryption is tangled up with all of this.  If we transmit data over
the global Internet, then generally we are going to use fieldz to 
prepare data for transmission and then encrypt the result before 
transmitting it as one large opaque block.

We also want to be able to use the fieldz code to serialize data
being written to disk.  For example XLattice in its current implementation
stores configuration for a node in XML and then deserializes it at boot
time.  What seems sensible is to replace xlattice.corexml with 
xlattice.fieldz and use that for serialization and deserialization.
That is an ambitious objective.


CHANGES
~/dev/py/fieldz/CHANGES

v0.11.19
    2018-03-11
        * edit setup.py and tox config files                        SLOC 5645
v0.11.18
    2018-01-08
        * 12 failed, 35 passed                                      SLOC 5646
v0.11.17
    2018-01-07
        * 13 failed, 34 passed                                      SLOC 5655
v0.11.16
    2018-01-06
        * 15 failed, 32 passed                                      SLOC 5649
v0.11.15
    2018-01-05
        * 16 failed, 31 passed                                      SLOC 5420
v0.11.14
    2018-01-04
        * 20 failed, 26 passed                                      SLOC 5405
v0.11.13
    2017-12-30
        * pylinting and debugging                                   SLOC 5395
v0.11.12
    2017-11-24
        *
v0.11.11
    2017-10-24
        * add projConfig.toml                                       SLOC 5401
v0.11.10
    2017-08-12
        * add tox.ini, *requirements.txt; edit cfg files            SLOC 5371
v0.11.9
    2017-07-15
        * rewrite MANIFEST.in, setup.py; minor edits config files   SLOC 5341
v0.11.8
    2017-06-09
        * restructure for pytest                                    SLOC 5367
v0.11.7
    2017-04-22
        * drop dist/* from repo; plus minor edits                   SLOC 6004
v0.11.6
    2017-02-06
        * more pylinting, more minor bug fixes                      SLOC 5689
v0.11.5
    2017-02-03
        * pylinting, minor bug fixes                                SLOC 5723
v0.11.4
    2017-02-01
        * test_{reg,log_entry}.py succeed                           SLOC 5702
v0.11.3
    2017-01-31
        * test_{big_test,enum,field_impl,gen_enum}.py succeed       SLOC 5635
v0.11.2
    2017-01-21
        * Quants and CoreTypes now standard enums                   SLOC 5646
    2017-01-16
        * renamed fieldz/reg/__init__.py fieldz/reg.py because of
            a weirdness in the python interpreter: it sometimes 
            cannot find a subdirectory when importing a
            constituent module. 
v0.11.1
    2017-01-15
        * introduced simpler Quants enum, dropped redundant code    SLOC 5579
v0.11.0
    2016-12-17
        * wireops now separate repo                                 SLOC 5550
v0.10.30
    2016-12-14
        * preparing to split off wireops as separate repo           SLOC 6911
v0.10.29
    2016-12-04
        * more pylinting                                            SLOC 7207
v0.10.28
    2016-11-04
        * 14 out of 22 tests succeed (8 FAILED)                     SLOC 7147
v0.10.27
    2016-10-31
        * INSTALLED BUT NOT PUSHED                                  SLOC 7120
v0.10.26
    2016-10-31
        * 11 out of 22 tests succeed                                SLOC 7115
v0.10.25
    2016-10-30
        * CoreTypes work, but should use metaclass                  SLOC 7093
v0.10.24
    2016-10-29
        * rename vars per PEP 8, phase 2                            sLOC 7081
v0.10.23
    2016-10-28
        * rename vars per PEP 8 (phase 1), some debugging           SLOC 7038
v0.10.22
    2016-10-27
        * rename API per PEP 8, with some debugging                 SLOC 6944
v0.10.21
    2016-10-17
        * rename files per PEP 8; add pylint and coverage           SLOC 6927
v0.10.20
    2016-08-24
        * fix license, setup.py                                     SLOC 6790
v0.10.19
    2016-08-13
        * elaborate README.md                                       SLOC 6635
v0.10.18
    2016-08-11
        * testFieldImpl.py succeeds                                 SLOC 6618
v0.10.17
    2016-08-09
        * <just done for alertz, to get current code>
v0.10.16
    2016-08-08
        * testReg.py succeeds                                       SLOC 6589
v0.10.15
    2016-08-05
        * 18/22 tests succeed                                       SLOC 6546
v0.10.14
    2016-08-03
        * error count halved?                                       SLOC 6538
v0.10.13
    2016-08-01
        * fixed MsgSpec.__init__() bugs                             SLOC 6539
v0.10.12
    2016-06-21
        * add ghpDoc/{css,img}/                                     SLOC 6333
v0.10.11
    2016-05-19
        * add bkp2U, updadte .dvcz                                  SLOC 5772
v0.10.10
    2016-04-25
        * fix countem; add .dvcz and projCfg{.,local}               SLOC 5772
v0.10.9
    2016-03-30
        * testFieldImpl.py succeeds                                 SLOC 5864
v0.10.8
    2016-03-29
        * add testFieldImpl.py, fieldImpl.py                        SLOC 5776
v0.10.7
    2016-03-28
        * down to two bugs (property, mutable constant)             SLOC 5720
            - where bug = *known* bug, a test failure
v0.10.6
    2016-03-19
        * fix testProtoSpec; add and execute auto8it                SLOC 5736
v0.10.5
    2016-03-17
        * testCoreType, testLogEntry fixed                          SLOC 5569
v0.10.4
    2016-02-23
        * fixed dead links, added Project Status                    SLOC 5562
v0.10.3
    2016-02-22
        * minor edits to code; bring in implementation notes        SLOC 5562
v0.10.2
    2016-02-20
        * added module names; edits to msgImply.py                  SLOC 5562
v0.10.1
    2016-02-19
        * 18/24 test*.py succeed                                    SLOC 5553
v0.10.0
    2016-02-16
        * converted to Python3, made ready for github               SLOC 4801
    2016-02-17
        * corrections to countem                                    SLOC 5552
v0.9.2
    2012-12-24
        * added vuInt32 length field in logEntry in examples
    2012-12-26
        * dropped '0xff&' from writeRawB160 serialization           SLOC 4814
v0.9.1
    2012-12-19
        * MSG_NAME.FIELD_NAME now returns value of field,           SLOC 4793
            at least for single-valued Q_REQUIRED fields
v0.9.0
    2012-12-19
        * MetaMessage renamed MetaMsg for consistency               SLOC 4799
v0.8.0  
    2012-12-18

fieldz/raw.py
# fieldz/raw.py

import ctypes
from fieldz.field_types import FieldTypes

# for debugging
#import binascii

__all__ = [
    'VARINT_TYPE', 'PACKED_VARINT_TYPE',
    'B32_TYPE', 'B64_TYPE', 'LEN_PLUS_TYPE',
    'B128_TYPE', 'B160_TYPE', 'B256_TYPE',

    'field_hdr', 'read_field_hdr', 'field_hdr_len',
    'hdr_field_nbr', 'hdr_type',
    'length_as_varint', 'write_varint_field',
    'read_raw_varint', 'write_raw_varint',
    'read_raw_b32', 'write_b32_field',
    'read_raw_b64', 'write_b64_field',
    'read_raw_len_plus', 'write_len_plus_field',
    'read_raw_b128', 'write_b128_field',
    'read_raw_b160', 'write_b160_field',
    'read_raw_b256', 'write_b256_field',
    # -- methods --------------------------------------------
    'next_power_of_two',
    # -- classes --------------------------------------------
    'WireBuffer',
]

# these are PRIMITIVE types, which determine the number of bytes
# occupied in the buffer; they are NOT data types
VARINT_TYPE = 0  # variable length integer
PACKED_VARINT_TYPE = 1  # variable length integer
B32_TYPE = 2  # fixed length, 32 bits
B64_TYPE = 3  # fixed length, 64 bits
LEN_PLUS_TYPE = 4  # sequence of bytes preceded by a varint length
B128_TYPE = 5  # fixed length, 128 bits (AES IV length)
B160_TYPE = 6  # fixed length, 160 bits (SHA1 content key)
B256_TYPE = 7  # fixed length, 128 bits (SHA3 content key)

# FIELD HEADERS #####################################################


def field_hdr(ndx, tstamp):
    # it would be prudent but slower to validate the parameters
    # DEBUG
    #   print "ndx = %u, t = %u, header is 0x%x" % (n, t, (n << 3) | t)
    # END
    return (ndx << 3) | tstamp


def field_hdr_len(nnn, tstamp):
    return length_as_varint(field_hdr(nnn, tstamp))


def hdr_field_nbr(ndx):
    return ndx >> 3


def hdr_type(ndx):
    return ndx & 7


def read_field_hdr(chan):
    hdr = read_raw_varint(chan)
    p_type = hdr_type(hdr)      # this is the primitive field type
    field_nbr = hdr_field_nbr(hdr)
    return (p_type, field_nbr)

# VARINTS ###########################################################


def length_as_varint(varint_):
    """
    Return the number of bytes occupied by an unsigned int.
    caller is responsible for assuring that v is in fact properly
    cast as unsigned occupying no more space than an int64 (and
    so no more than 10 bytes).
    """
    if varint_ < (1 << 7):
        return 1
    elif varint_ < (1 << 14):
        return 2
    elif varint_ < (1 << 21):
        return 3
    elif varint_ < (1 << 28):
        return 4
    elif varint_ < (1 << 35):
        return 5
    elif varint_ < (1 << 42):
        return 6
    elif varint_ < (1 << 49):
        return 7
    elif varint_ < (1 << 56):
        return 8
    elif varint_ < (1 << 63):
        return 9
    else:
        return 10


def read_raw_varint(chan):
    buf = chan.buffer
    offset = chan.position
    varint_ = 0
    ndx_ = 0
    while True:
        if offset >= len(buf):
            raise ValueError("attempt to read beyond end of buffer")
        next_byte = buf[offset]
        offset += 1

        sign = next_byte & 0x80
        next_byte = next_byte & 0x7f
        next_byte <<= (ndx_ * 7)
        varint_ |= next_byte
        ndx_ += 1

        if sign == 0:
            break
    chan.position = offset
    return varint_


def write_raw_varint(chan, string):
    buf = chan.buffer
    offset = chan.position
    # all varints are construed as 64 bit unsigned numbers
    varint_ = ctypes.c_uint64(string).value
#   # DEBUG
#   print "entering writeRaw: will write 0x%x at offset %u" % ( v, offset)
#   # END
    len_ = length_as_varint(varint_)
    if offset + len_ > len(buf):
        raise ValueError("can't fit varint of length %u into buffer" % len_)
    while True:
        buf[offset] = (varint_ & 0x7f)
        offset += 1
        varint_ >>= 7
        if varint_ == 0:
            chan.position = offset   # next unused byte
            break
        else:
            buf[offset - 1] |= 0x80


def write_varint_field(chan, varint_, nnn):
    # the header is the field number << 3 ORed with 0, VARINT_TYPE
    hdr = field_hdr(nnn, VARINT_TYPE)
    write_raw_varint(chan, hdr)
#   # DEBUG
#   print "header was 0x%x; writing value 0x%x at offset %u" % (
#                                               hdr, v, chan.position)
#   # END
    write_raw_varint(chan, varint_)

# 32- AND 64-BIT FIXED LENGTH FIELDS ################################


def read_raw_b32(chan):
    """ buf construed as array of unsigned bytes """
    buf = chan.buffer
    offset = chan.position
    # XXX verify buffer long enough
    varint_ = buf[offset]
    offset += 1         # little-endian
    varint_ |= buf[offset] << 8
    offset += 1
    varint_ |= buf[offset] << 16
    offset += 1
    varint_ |= buf[offset] << 24
    offset += 1
    chan.position = offset
    return varint_


def write_raw_b32(chan, varint_):
    buf = chan.buffer
    offset = chan.position
    buf[offset] = 0xff & varint_
    varint_ >>= 8
    offset += 1
    buf[offset] = 0xff & varint_
    varint_ >>= 8
    offset += 1
    buf[offset] = 0xff & varint_
    varint_ >>= 8
    offset += 1
    buf[offset] = 0xff & varint_
    offset += 1
    chan.position = offset


def write_b32_field(chan, varint_, file):
    hdr = field_hdr(file, B32_TYPE)
    write_raw_varint(chan, hdr)
    write_raw_b32(chan, varint_)


def read_raw_b64(chan):
    """ buf construed as array of unsigned bytes """
    buf = chan.buffer
    offset = chan.position
    # XXX verify buffer long enough
    varint_ = buf[offset]
    offset += 1         # little-endian
    varint_ |= buf[offset] << 8
    offset += 1
    varint_ |= buf[offset] << 16
    offset += 1
    varint_ |= buf[offset] << 24
    offset += 1
    varint_ |= buf[offset] << 32
    offset += 1
    varint_ |= buf[offset] << 40
    offset += 1
    varint_ |= buf[offset] << 48
    offset += 1
    varint_ |= buf[offset] << 56
    offset += 1
    chan.position = offset
    return varint_


def write_raw_b64(chan, varint_):
    # XXX verify buffer long enough
    buf = chan.buffer
    offset = chan.position
    buf[offset] = 0xff & varint_
    varint_ >>= 8
    offset += 1
    buf[offset] = 0xff & varint_
    varint_ >>= 8
    offset += 1
    buf[offset] = 0xff & varint_
    varint_ >>= 8
    offset += 1
    buf[offset] = 0xff & varint_
    varint_ >>= 8
    offset += 1
    buf[offset] = 0xff & varint_
    varint_ >>= 8
    offset += 1
    buf[offset] = 0xff & varint_
    varint_ >>= 8
    offset += 1
    buf[offset] = 0xff & varint_
    varint_ >>= 8
    offset += 1
    buf[offset] = 0xff & varint_
    offset += 1
    chan.position = offset


def write_b64_field(chan, varint_, file):
    hdr = field_hdr(file, B64_TYPE)
    write_raw_varint(chan, hdr)
    write_raw_b64(chan, varint_)

# VARIABLE LENGTH FIELDS ############################################


def read_raw_len_plus(chan):

    # read the varint len
    len_ = read_raw_varint(chan)
    buf = chan.buffer
    offset = chan.position

#   # DEBUG
#   print "readRawLenPlus: length of text is %d bytes" % len_
#   # END

    # then read n actual bytes
    string = []
    count = 0
    while count < len_:
        string.append(buf[offset])
        count += 1
        offset += 1
    chan.position = offset
    return bytearray(string)


def write_raw_bytes(chan, bytes_):
    """ bytes a byte array """
    buf = chan.buffer
    offset = chan.position
    # XXX CHECK LEN OFFSET

    # DEBUG
    print("writeRawBytes: type(bytes) is ", type(bytes_))
    # END

    for b_val in bytes_:
        buf[offset] = int(b_val)
        offset += 1
#   # DEBUG
    print("wrote '%s' as %u raw bytes" % (str(bytes_), len(bytes_)))
#   # END
    chan.position = offset

# XXX 2012-12-11 currently used only in one place


def write_field_hdr(chan, field_nbr, prim_type):
    """ write the field header """
    hdr = field_hdr(field_nbr, prim_type)
    write_raw_varint(chan, hdr)


def write_len_plus_field(chan, string, file):
    """s is a bytearray or string"""
    write_field_hdr(chan, file, LEN_PLUS_TYPE)
    # write the length of the byte array --------
    write_raw_varint(chan, len(string))

    # now write the byte array itself -----------
    write_raw_bytes(chan, string)

# LONGER FIXED-LENGTH BYTE FIELDS ===================================


def read_raw_b128(chan):
    """ buf construed as array of unsigned bytes """
    # XXX verify buffer long enough
    buf = chan.buffer
    offset = chan.position
    string = []
    for i in range(16):
        string.append(buf[offset + i])
    offset += 16
    chan.position = offset
    return bytearray(string)


def write_raw_b128(chan, varint_):
    """ v is a bytearray or string """
    buf = chan.buffer
    offset = chan.position
    for i in range(16):
        # this is a possibly unnecessary cast
        buf[offset] = 0xff & varint_[i]
        offset += 1
    chan.position = offset


def write_b128_field(chan, varint_, file):
    hdr = field_hdr(file, B128_TYPE)
    write_raw_varint(chan, hdr)
    write_raw_b128(chan, varint_)                  # GEEP


def read_raw_b160(chan):
    """ buf construed as array of unsigned bytes """
    # XXX verify buffer long enough
    buf = chan.buffer
    offset = chan.position
    string = []
    for i in range(20):
        string.append(buf[offset + i])
    offset += 20
    chan.position = offset
    return bytearray(string)


def write_raw_b160(chan, varint_):
    """ v is a bytearray or string """
    buf = chan.buffer
    offset = chan.position
    for i in range(20):
        buf[offset] = varint_[i]
        offset += 1
    chan.position = offset


def write_b160_field(chan, varint_, file):
    hdr = field_hdr(file, B160_TYPE)
    write_raw_varint(chan, hdr)
    write_raw_b160(chan, varint_)                  # GEEP


def read_raw_b256(chan):
    """ buf construed as array of unsigned bytes """
    # XXX verify buffer long enough
    buf = chan.buffer
    offset = chan.position
    string = []
    for i in range(32):
        string.append(buf[offset + i])
    offset += 32
    chan.position = offset
    return bytearray(string)


def write_raw_b256(chan, varint_):
    """ v is a bytearray or string """
    buf = chan.buffer
    offset = chan.position
    # DEBUG
    # print "DEBUG: writeRawB256 datum len is %s" % len(v)
    # END
    for i in range(32):
        # print "DEBUG:    v[%u] = %s" % (i, binascii.b2a_hex(v[i]))
        buf[offset] = varint_[i]
        offset += 1
    chan.position = offset


def write_b256_field(chan, varint_, file):
    hdr = field_hdr(file, B256_TYPE)
    write_raw_varint(chan, hdr)
    write_raw_b256(chan, varint_)                  # GEEP

# PRIMITIVE FIELD NAMES =============================================


class PrimFields(object):
    """ lower-level primitive field types """

    _P_VARINT = 0
    _P_B32 = 1     # 32 bit fields
    _P_B64 = 2     # 64 bit fields
    _P_LEN_PLUS = 3     # varint len followed by that many bytes
    # the following can be implemented in terms of _P_LEN_PLUS
    _P_B128 = 4    # fixed length string of 16 bytes
    _P_B160 = 5    # fixed length string of 20 bytes
    _P_B256 = 6    # fixed length string of 32 bytes

    _MAX_TYPE = _P_B256

    # none of these (pVarint..pB256) is currently used
#   @property
#   def pVarint(clz):       return clz._P_VARINT
#   @property
#   def pB32(clz):          return clz._P_B32
#   @property
#   def pB64(clz):          return clz._P_B64
#   @property
#   def pLenPlus(clz):      return clz._P_LenPlus
#   @property
#   def pB128(clz):         return clz._P_B128
#   @property
#   def pB160(clz):         return clz._P_B160
#   @property
#   def pB256(clz):         return clz._P_B256

    _names = {}
    _names[_P_VARINT] = 'pVarint'
    _names[_P_B32] = 'pB32'
    _names[_P_B64] = 'pB64'
    _names[_P_LEN_PLUS] = 'pLenPlus'
    _names[_P_B128] = 'pB128'
    _names[_P_B160] = 'pB160'
    _names[_P_B256] = 'pB256'

    @classmethod
    def name(cls, varint_):
        if varint_ is None or varint_ < 0 or FieldTypes._MAX_TYPE < varint_:
            raise ValueError('no such field type: %s' + str(varint_))
        return cls._names[varint_]

# -- WireBuffer -----------------------------------------------------


def next_power_of_two(nnn):
    """
    If n is a power of two, return n.  Otherwise return the next
    higher power of 2.
    See eg http://acius2.blogspot.com/2007/11/calculating-next-power-of-2.html
    """
    if nnn < 1:
        raise ValueError("nextPowerOfTwo: %s < 1" % str(nnn))
    nnn = nnn - 1
    nnn = (nnn >> 1) | nnn
    nnn = (nnn >> 2) | nnn
    nnn = (nnn >> 4) | nnn
    nnn = (nnn >> 8) | nnn
    nnn = (nnn >> 16) | nnn
    return nnn + 1


class WireBuffer(object):

    __slots__ = ['_buffer', '_capacity', '_limit', '_position', ]

    def __init__(self, nnn=1024, buffer=None):
        """
        Initialize the object.  If a buffer is specified, use it.
        Otherwise create one.  The resulting buffer will have a
        capacity which is a power of 2.
        """
        if buffer:
            self._buffer = buffer
            buf_size = len(buffer)
            if nnn < buf_size:
                nnn = buf_size
            nnn = next_power_of_two(nnn)
            # DANGER: buffer capacities of cloned buffers can get out of sync
            if nnn > buf_size:
                more = bytearray(nnn - buf_size)
                self.buffer.extend(more)
        else:
            nnn = next_power_of_two(nnn)
            # allocate and initialize the buffer; init probably a waste of time
            self._buffer = bytearray(nnn)

        self._capacity = nnn
        self._limit = nnn
        self._position = 0

    def copy(self):
        """
        Returns a copy of this WireBuffer using the same underlying
        bytearray.
        """
        return WireBuffer(len(self._buffer), self._buffer)

    @property
    def buffer(self):
        return self._buffer

    @property
    def position(self):
        return self._position

    @position.setter
    def position(self, offset):
        if offset < 0:
            raise ValueError('position cannot be negative')
        if offset >= len(self._buffer):
            raise ValueError('position cannot be beyond capacity')
        self._position = offset

    @property
    def limit(self):
        return self._limit

    @limit.setter
    def limit(self, offset):
        if offset < 0:
            raise ValueError('limit cannot be set to a negative')
        if offset < self._position:
            raise ValueError(
                "limit can't be set to less than current position")
        if offset > self._capacity:
            raise ValueError('limit cannot be beyond capacity')
        self._limit = offset

    @property
    def capacity(self):
        return len(self._buffer)

    def reserve(self, k):
        """
        We need to add k more bytes; if the buffer isn't big enough,
        resize it.
        """
        if k < 0:
            raise ValueError(
                "attempt to increase WireBuffer size by negative number of bytes")
        if self._position + k >= self._capacity:
            # wildly inefficient, I'm sure
            more = bytearray(self._capacity)
            self._buffer.extend(more)
            self._capacity *= 2
